tests/sentry/api/endpoints/test_organization_details.py

```diff
@@ -764,11 +766,69 @@ def test_update_name_with_mapping(self):
         org = Organization.objects.get(id=organization_id)
         assert org.name == "SaNtRy"
 
+        with outbox_runner():
+            pass
+
         with exempt_from_silo_limits():
             assert OrganizationMapping.objects.filter(
                 organization_id=organization_id, name="SaNtRy"
             ).exists()
 
+    def test_update_slug(self):
+        organization_mapping = OrganizationMapping.objects.get(organization_id=self.organization.id)
+        assert organization_mapping.slug == self.organization.slug
+
+        desired_slug = "new-santry"
+        self.get_success_response(self.organization.slug, slug=desired_slug)
+        self.organization.refresh_from_db()
+        assert self.organization.slug == desired_slug
+
+        # Ensure that the organization update has been flushed
+        with outbox_runner():
+            pass
+
+        organization_mapping.refresh_from_db()
+        assert organization_mapping.slug == desired_slug
+
+    def test_update_slug_with_temporary_rename_collision(self):
+        desired_slug = "taken"
+        previous_slug = self.organization.slug
+        org_with_colliding_slug = self.create_organization(
+            slug=desired_slug, name="collision-imminent"
+        )
+
+        # Drain the initial slug creation to ensure a mapping exists for the new org
+        Organization.outbox_for_update(org_id=org_with_colliding_slug.id).drain_shard()
+        colliding_org_mapping = OrganizationMapping.objects.get(
+            organization_id=org_with_colliding_slug.id
+        )
+        assert colliding_org_mapping.slug == desired_slug
+
+        # Queue a slug update but don't drain the shard yet to ensure a temporary collision happens
+        org_with_colliding_slug.slug = "unique-slug"
+        org_with_colliding_slug.save()
+
+        self.get_success_response(self.organization.slug, slug=desired_slug)
+        self.organization.refresh_from_db()
+        assert self.organization.slug == desired_slug
+
+        # Ensure that the organization update has been flushed, but it collides when attempting an upsert
+        with pytest.raises(IntegrityError):
+            Organization.outbox_for_update(org_id=self.organization.id).drain_shard()
+
+        organization_mapping = OrganizationMapping.objects.get(organization_id=self.organization.id)
+        assert organization_mapping.slug == previous_slug
+
+        # Flush the colliding org slug change
+        Organization.outbox_for_update(org_id=org_with_colliding_slug.id).drain_shard()
+        colliding_org_mapping.refresh_from_db()
+        assert colliding_org_mapping.slug == "unique-slug"
+
+        # Flush the desired slug change and assert the correct slug was resolved
+        Organization.outbox_for_update(org_id=self.organization.id).drain_shard()
+        organization_mapping.refresh_from_db()
+        assert organization_mapping.slug == desired_slug
```

Comment:
I'm really not sure if this test belongs here or elsewhere, but I wanted to test the race-condition surrounding slug rename collisions somehow 🤔